### ISSUE TITLE
How to obtain validation loss during training?

### ISSUE THREAD

**Issue by BAOUCH** [filaPro/oneformer3d#118]
Hello, thank you for your great work on OneFormer3D!

I am currently training the model using the provided configuration, and I can see the training loss being logged. However, I do not see the validation loss in the logs or evaluation outputs.

- Is the validation loss computed at all during training?

- If so, where can I find it or enable its logging?

- If not, what would be the best way to modify the config or runner to include validation loss?

Thanks in advance for your help!

**Comment by filaPro** [comment]
Loss is not computed during validation. So you can compute as a a metric and log [here](https://github.com/filaPro/oneformer3d/blob/main/oneformer3d/unified_metric.py#L216).
It's not that easy as requires copying the whole logic from [criterion](https://github.com/filaPro/oneformer3d/blob/main/oneformer3d/unified_criterion.py#L6) to metric.

Another option is to modify `test_pipeline` in config to have the same gt [keys](https://github.com/filaPro/oneformer3d/blob/main/configs/oneformer3d_1xb4_scannet.py#L164) as `train_pipeline` [keys](https://github.com/filaPro/oneformer3d/blob/main/configs/oneformer3d_1xb4_scannet.py#L128). Then modify [predict](https://github.com/filaPro/oneformer3d/blob/main/oneformer3d/oneformer3d.py#L375) function to accept these new keys and return loss same as [loss](https://github.com/filaPro/oneformer3d/blob/main/oneformer3d/oneformer3d.py#L335) function.

**Comment by BAOUCH** [comment]
Thank you for the response!

I've implemented the second option you suggested and wanted to share the results and modifications I made (hopefully helpful to others too!).

Changes Made:
**1. Modified test_pipeline Configuration:**
```
test_pipeline = [
    dict(
        type='LoadPointsFromFile',
        coord_type='DEPTH',
        shift_height=False,
        use_color=True,
        load_dim=6,
        use_dim=[0, 1, 2, 3, 4, 5]),
    dict(
        type='LoadAnnotations3D',
        with_bbox_3d=False,
        with_label_3d=False,
        with_mask_3d=True,
        with_seg_3d=True),
    dict(type='PointInstClassMapping_', 
        num_classes=num_instance_classes),
    dict(
        type='MultiScaleFlipAug3D',
        img_scale=(1333, 800),
        pts_scale_ratio=1,
        flip=False,
        transforms=[
            dict(
                type='NormalizePointsColor_',
                color_mean=[127.5, 127.5, 127.5])]),
    dict(type='Pack3DDetInputs_', keys=['points', 'gt_labels_3d', 'pts_semantic_mask', 'pts_instance_mask'])
```

**2. Enhanced predict Function in oneformer3d.py:**
Modified the predict function to calculate validation loss when ground truth is available:
```
def predict(self, batch_inputs_dict, batch_data_samples, **kwargs):
    coordinates, features, inverse_mapping, spatial_shape = self.collate(
        batch_inputs_dict['points'])
    x = spconv.SparseConvTensor(
        features, coordinates, spatial_shape, len(batch_data_samples))

    x = self.extract_feat(x)
    x = self.decoder(x)

    # Check if ground truth is available
    has_gt = all(hasattr(data_sample, 'gt_pts_seg') and 
                 hasattr(data_sample.gt_pts_seg, 'pts_instance_mask') and
                 hasattr(data_sample.gt_pts_seg, 'pts_semantic_mask')
                 for data_sample in batch_data_samples)
    
    if has_gt:
        sp_gt_instances = []
        for i in range(len(batch_data_samples)):
            voxel_superpoints = inverse_mapping[coordinates[:, 0][inverse_mapping] == i]
            voxel_superpoints = torch.unique(voxel_superpoints, return_inverse=True)[1]
            inst_mask = batch_data_samples[i].gt_pts_seg.pts_instance_mask
            sem_mask = batch_data_samples[i].gt_pts_seg.pts_semantic_mask

            from types import SimpleNamespace
            gt_instances = SimpleNamespace()
            gt_instances.sp_sem_masks = self.get_gt_semantic_masks(
                sem_mask, voxel_superpoints, self.num_classes)
            gt_instances.sp_inst_masks = self.get_gt_inst_masks(
                inst_mask, voxel_superpoints)
            sp_gt_instances.append(gt_instances)

        validation_loss = self.criterion(x, sp_gt_instances)
        self._accumulate_validation_loss(validation_loss)

    results_list = self.predict_by_feat(x, inverse_mapping)

    for i, data_sample in enumerate(batch_data_samples):
        data_sample.pred_pts_seg = results_list[i]

    return batch_data_samples
```

**3. Added Loss Accumulation and Logging Methods:**

```
    def _accumulate_validation_loss(self, validation_loss):
        """Accumulate validation loss for epoch-level averaging.
        
        Args:
            validation_loss (dict): Dictionary containing loss components.
        """
        # Initialize accumulator if it doesn't exist
        if not hasattr(self, '_val_loss_accumulator'):
            self._val_loss_accumulator = {}
            self._val_loss_count = 0
        
        # Accumulate losses
        for key, value in validation_loss.items():
            loss_value = value.item() if hasattr(value, 'item') else float(value)
            if key not in self._val_loss_accumulator:
                self._val_loss_accumulator[key] = 0.0
            self._val_loss_accumulator[key] += loss_value
        
        self._val_loss_count += 1

    def log_epoch_validation_loss(self, epoch):
        """Log the average validation loss for the current epoch and reset accumulator.
        
        Args:
            epoch (int): Current epoch number.
        """
        if not hasattr(self, '_val_loss_accumulator') or self._val_loss_count == 0:
            return
        
        import os
        import logging
        from datetime import datetime
        
        # Create logs directory if it doesn't exist
        log_dir = "validation_logs"
        os.makedirs(log_dir, exist_ok=True)
        
        # Setup logger for validation loss
        log_file = os.path.join(log_dir, "epoch_validation_loss.log")
        
        # Create a unique logger for validation to avoid conflicts
        logger_name = f"epoch_validation_loss_{id(self)}"
        logger = logging.getLogger(logger_name)
        
        # Only add handler if it doesn't exist
        if not logger.handlers:
            logger.setLevel(logging.INFO)
            handler = logging.FileHandler(log_file, mode='a')
            formatter = logging.Formatter('%(asctime)s - %(message)s')
            handler.setFormatter(formatter)
            logger.addHandler(handler)
            logger.propagate = False  # Prevent double logging
        
        # Calculate average losses
        avg_losses = {}
        for key, total_loss in self._val_loss_accumulator.items():
            avg_losses[key] = total_loss / self._val_loss_count
        
        # Format loss information
        timestamp = datetime.now().strftime("%Y-%m-%d %H:%M:%S")
        loss_info = []
        
        for key, avg_loss in avg_losses.items():
            loss_info.append(f"{key}: {avg_loss:.6f}")
        
        loss_str = " | ".join(loss_info)
        

        
        # Log the epoch validation loss
        log_message = f"EPOCH {epoch} VALIDATION_LOSS (avg over {self._val_loss_count} batches) - {loss_str}"
        logger.info(log_message)
        
        # Also print to console for immediate feedback
        print(f"[{timestamp}] {log_message}")
        
        # Reset accumulator for next epoch
        self._val_loss_accumulator = {}
        self._val_loss_count = 0

    def reset_validation_accumulator(self):
        """Reset validation loss accumulator. Call this at the start of each epoch."""
        self._val_loss_accumulator = {}
        self._val_loss_count = 0

```

**4. Custom Validation Loss Hook**
Created validation_loss_hook.py:
```
from mmengine.hooks import Hook
from mmdet3d.registry import HOOKS

@HOOKS.register_module()
class ValidationLossHook(Hook):
    """Custom hook to log validation loss at the end of each epoch."""
    
    def __init__(self, priority='NORMAL'):
        self.priority = priority
    
    def before_val_epoch(self, runner):
        """Called before validation epoch starts."""
        model = runner.model
        if hasattr(model, 'reset_validation_accumulator'):
            model.reset_validation_accumulator()
    
    def after_val_epoch(self, runner, metrics=None):
        """Called after validation epoch ends."""
        model = runner.model
        if hasattr(model, 'log_epoch_validation_loss'):
            current_epoch = runner.epoch
            model.log_epoch_validation_loss(current_epoch)
```
And added it to the config:
```
custom_hooks = [
    dict(type='EmptyCacheHook', after_iter=True),
    dict(type='ValidationLossHook')  # Add the custom hook
]
```

**5. Results**

The implementation is working, Here are the validation loss results from my test run (Saved in epoch_validation_loss.log) (Note: I used reduced hyperparameters and only ran 3 epochs for debugging.):
```
2025-05-27 00:50:12,638 - EPOCH 1 VALIDATION_LOSS (avg over 68 batches) - inst_loss: 14.095709 | last_layer_seg_loss: 21.133884 | layer_0_seg_loss: 12.316557 | layer_1_seg_loss: 17.133181 | layer_2_seg_loss: 18.688017

2025-05-27 00:52:04,400 - EPOCH 2 VALIDATION_LOSS (avg over 68 batches) - inst_loss: 11.601339 | last_layer_seg_loss: 4.460294 | layer_0_seg_loss: 3.730294 | layer_1_seg_loss: 5.206560 | layer_2_seg_loss: 4.654749

2025-05-27 00:53:50,759 - EPOCH 3 VALIDATION_LOSS (avg over 68 batches) - inst_loss: 10.830762 | last_layer_seg_loss: 3.315537 | layer_0_seg_loss: 3.279136 | layer_1_seg_loss: 3.841044 | layer_2_seg_loss: 3.578979
```

**6. question:**

Right now, the loss dictionary returned by self.criterion() includes:

- inst_loss
- last_layer_seg_loss
- layer_0_seg_loss
- layer_1_seg_loss
- layer_2_seg_loss

Could you kindly explain where i can exactly find the "loss" metric?, like the one that is calculated in the train epochs, for example:
```
{"lr": 9.94714688307272e-05, "data_time": 0.021494174003601076, "grad_norm": 11.736356163024903, 👉"loss": 6.346024580001831👈, "inst_loss": 4.832869300842285, "last_layer_seg_loss": 0.3557536840438843, "layer_0_seg_loss": 0.43527495980262754, "layer_1_seg_loss": 0.37255752593278885, "layer_2_seg_loss": 0.34956909596920016, "time": 0.2226740837097168, "epoch": 4, "memory": 1451, "step": 406}
```

**Comment by filaPro** [comment]
Isn't this loss a sum of other mentioned losses?

**Comment by BAOUCH** [comment]
You are right, what a dumb question to ask 😅. Thanks a lot for your help!

**Comment by BAOUCH** [comment]
Sorry for the disturbance once again, but I have a question: why weren't you interested in calculating the validation loss during your work on OneFormer3D to detect possible overfitting or underfitting of the data?

**Comment by filaPro** [comment]
I think we log mAP metric during validation. Isn't it even better to detect over fitting?
